Extend model construction with assignment exclusion set

ajreynol opened the pull request:
This extends the core model building algorithm in CVC4 with "assignment exclusion sets".  This functionality is useful for assigning values to terms of finite type that have specific restrictions on what their value cannot be.

In detail, previously, all unassigned terms of finite type were always assigned the first term in the type enumeration.  This is done since it is assumed that theories (e.g. datatypes) always do enough work to ensure that *arbitrary* values can be assigned to terms of finite type, and *fresh* values can be assigned to terms of infinite type. However, there are compelling cases (sets+cardinality for finite element types) where one may want to impose restrictions on what values can be assigned to terms of finite types.  Thus, we now provide `setAssignmentExclusionSet` as a way of communicating these restrictions. 

This PR also refactors a few parts of `TheoryEngineModelBuilder::buildModel` to be clearer, in particular by adding a few helper functions, and by caching certain information early in the function instead of recomputing it.

This is work towards #1123.

barrettcw commented on src/theory/theory_model_builder.cpp:
```diff
@@ -436,6 +526,148 @@ bool TheoryEngineModelBuilder::buildModel(Model* m)
     }
   }
 
+  Trace("model-builder") << "Compute assignable information..." << std::endl;
```
The buildModel function is getting really big.  Can we break this new code out into a separate function?

barrettcw commented on src/theory/theory_model_builder.cpp:
```diff
@@ -436,6 +526,148 @@ bool TheoryEngineModelBuilder::buildModel(Model* m)
+    eqcs_i = eq::EqClassesIterator(ee);
+    bool assignable = false;
+    bool evaluable = false;
+    for (; !eqcs_i.isFinished(); ++eqcs_i)
```
Please document this loop - something like "loop through all of the equivalence classes doing X with each of them"

barrettcw commented on src/theory/theory_model_builder.cpp:
```diff
@@ -436,6 +526,148 @@ bool TheoryEngineModelBuilder::buildModel(Model* m)
+      bool hasESet = false;
+      bool foundESet = false;
+      eq::EqClassIterator eqc_i = eq::EqClassIterator(eqc, ee);
+      for (; !eqc_i.isFinished(); ++eqc_i)
```
Document this loop - in particular what are the loop exit conditions

barrettcw commented on src/theory/theory_model_builder.cpp:
```diff
@@ -436,6 +526,148 @@ bool TheoryEngineModelBuilder::buildModel(Model* m)
+      // for assignment exclusion sets
+      std::vector<Node> group;
+      std::vector<Node> eset;
+      bool hasESet = false;
```
Document this variable

barrettcw commented on src/theory/theory_model_builder.cpp:
```diff
@@ -436,6 +526,148 @@ bool TheoryEngineModelBuilder::buildModel(Model* m)
+      std::vector<Node> group;
+      std::vector<Node> eset;
+      bool hasESet = false;
+      bool foundESet = false;
```
Same - what is the difference between hasESet and foundESet

barrettcw commented on src/theory/theory_model_builder.cpp:
```diff
@@ -436,6 +526,148 @@ bool TheoryEngineModelBuilder::buildModel(Model* m)
+    std::unordered_set<Node, NodeHashFunction> processed;
+    eqcs_i = eq::EqClassesIterator(ee);
+    bool assignable = false;
+    bool evaluable = false;
```
Document these variables.  There seem to be several invariants connected with each of them.

ajreynol commented on src/theory/theory_model.h:
```diff
@@ -299,9 +353,22 @@ class TheoryModel : public Model
   std::map<Node, Node> d_reps;
+  /** Map of terms to their assignment exclusion set. */
+  std::map<Node, std::vector<Node> > d_assignExcSet;
+  /**
```
std::map tends to be faster than std::unordered_map in my experience when the number of elements you are storing is small (<15).  This map is expected to be very small (it is at most the # of types T such that (Set T) exists currently).  The other I suppose should be changed.

ajreynol commented:
I've addressed the review. Notice I managed to extract out that block of code out of `buildModel` to a new function `computeAssignableInfo` without too much headache.

barrettcw commented on src/theory/theory_model_builder.h:
```diff
@@ -253,7 +253,37 @@ class TheoryEngineModelBuilder : public ModelBuilder
+   * require special ways of assigning values, e.g. those that take into
+   * account assignment exclusion sets,
+   * (4) eqcToAssignerMaster: a map from equivalence classes to the equivalence
+   * class that it shares an assigner object (all elements in the range of this
```
"shares an assigner object" -> "shares an assigner object with"

ajreynol commented on src/theory/theory_model_builder.cpp:
```diff
@@ -26,10 +26,99 @@ using namespace CVC4::context;
+  }
+  std::vector<Node>& eset = a.d_assignExcSet;
+  std::map<Node, Node>::iterator it;
+  for (unsigned i = 0, size = eset.size(); i < size; i++)
```
We can't since we update at an index on line 116.

4tXJ7f replied:
It should be possible to just do it with `for (Node& er : eset)`.

ajreynol replied:
This is pretty unintuitive to me. I'm actually surprised this works.  Since Node is immutable, I'm not sure why modifying the `er` reference in this case would modify `eset`?

4tXJ7f replied:
It works because it is a reference to the element in the `eset` vector (and not a copy of it). So assigning something to it is equivalent to `eset[i] = ...`. Since we are not really using references to `Node`s, maybe we should just leave it as it is though.

ajreynol commented on src/theory/theory_model_builder.h:
```diff
@@ -210,6 +215,75 @@ class TheoryEngineModelBuilder : public ModelBuilder
+     */
+    void initialize(TypeNode tn,
+                    TypeEnumeratorProperties* tep,
+                    const std::vector<Node>& aes);
```
The issue is that this class is being used as (non-pointer) data in a range of a map, so its construction is done by accessing the map via:
`      Assigner& a = eqcToAssigner[eqc];`
My understand of C++ is that the default constructor will be run for data in range of maps and hence we'd end up with an extra invocation of a constructor if we overwrote the data in the map by the non-default version e.g.:
`      Assigner& a = Assigner( ... );  eqcToAssigner[eqc] = a;`
where notice `eqcToAssigner[eqc] ` constructs a 2nd object.  Or perhaps my understanding is not correct.

This is generally the reason why I introduce "initialize" methods, I do agree constructors when possible is cleaner.

4tXJ7f replied:
Yes, `eqcToAssigner[eqc]` will create an instance using the default constructor if `eqc` is not already in the map. However, if you know that something is already in the map, you can use [`map::at()`](https://en.cppreference.com/w/cpp/container/map/at) and to add a new element, you can use [`map::emplace()`](https://en.cppreference.com/w/cpp/container/map/emplace). If you use these methods, the class shouldn't need a default constructor.

ajreynol replied:
Emplace is not a part of my vocabulary :)  

Generally I have a strong preference for keeping the set of primitive C++ language features I use to a minimum. While the `at`/`emplace` or similar feature might do the trick, they make the code less readable in my opinion.

However, generally I agree with you that we should use constructors instead of initialize methods. In this case however, I prefer keeping this code as is since it is a single-use class that is private to TheoryModelBuilder.

4tXJ7f replied:
The `emplace*` methods in the STL are actually quite useful because they create objects in place (so no need for a move/copy), so they tend to be more efficient. Instead of using `at()`, using `find()` would also do the trick. Since the class is local, I am ok with leaving it as-is, though.

4tXJ7f commented on src/theory/theory_model_builder.cpp:
```diff
@@ -26,10 +26,99 @@ using namespace CVC4::context;
+  // check if we have run out of elements
+  if (te.isFinished())
+  {
+    Assert(false);
```
I was actually referring to the comment in the header file :) We should also make it explicit there that this should never happen.

4tXJ7f commented on src/theory/theory_model.cpp:
```diff
@@ -526,6 +526,8 @@ void TheoryModel::assertSkeleton(TNode n)
                                             const std::vector<Node>& eset)
 {
+  // should not be assigned yet
+  Assert(d_assignExcSet.find(n)==d_assignExcSet.end());
```
Apply ClangFormat.
